_drafts/offensive-programming.md

```diff
@@ -0,0 +1,562 @@
+  specs, the system will immediately alert the developer. It's better than
+  tracking down what `-3` or `known_error` means and how it was bubbled up
+  through the system.
+- **Awareness of prevalence** - No developer or customer notices bugs unless the
+  bug is obvious. By forcing the software to reset, awareness of these undefined
+  bugs is increased and developers can now understand how often they are
+  occurring.
+
+## Offensive Programming in Practice
+
+Let's dig into some concrete examples about bugs we can find by using Offensive
+Programming practices.
+
+### Argument Validation
+
+If a developer comes along and tries to use an API and passes invalid arguments
+into the function, make sure the application yells at them to fix the issue.
+There is nothing worse than receiving a `-1` return value and digging through 10
+layers of firmware code only to find out the real reason was that you passed in
+a string of 20 characters when the maximum was 16 or because you accidentally
+passed a null pointer due to an uninitialized variable.
+
+```c
+void device_set_name(char *name, size_t name_len) {
+  ASSERT(name && name_len =< 16);
+  ...
+}
+```
+
+The only time I would choose, or at least heavily debate, to not use asserts to
+validate arguments is when I am building a library that will be used by people
+outside of my organization. In those cases, I would make the validation asserts
+optional, just as FreeRTOS does with their RTOS functions by allowing developers
+to define ``configASSERT` themselves[^freertos_asserts].
+
+### Resource Depletion
+
+Although using dynamic memory in embedded systems is sometimes frowned upon, it
+is often necessary to use it on complex systems that don't have enough static
+memory to go around. Even when dynamic memory is used, running out of memory
+should rarely occur. When memory is low, systems should adapt and put back
+pressure on any data flowing into the system and rate-limit memory-intensive
+operations.
+
+However, if a firmware _does_ deplete the dynamic memory pool, we want to know
+when and why it happened! It could be a memory leak or an accidental allocation
+that consumed most of the heap. Running out of memory might not be a
+show-stopping issue because the system _might_ recover, but if we are in the
+development or internal testing phase, we need to investigate this issue.
+
+To do this, we can add an assert inside of a `malloc()` function that verifies
+that the call did not fail.
+
+```c
+void *malloc_assert(size_t n) {
+  void *p = malloc(n)
+  ASSERT(p);
+  return p;
+  ...
+}
+```
+
+In code where the calls to `malloc()` should never fail, such as RTOS structure
+allocation, request and response buffers, etc., we can use this asserted
+version.
+
+Another common issue with RTOS-based systems is a queue becoming full and
+possibly not emptying quickly enough. As with memory depletion, this isn't a
+show-stopping issue, because the system might recover, but we want to look into
```

Comment:
Could call out this means we are dropping events which could be important/critical to operation

Reply:
Great point